### ISSUE TITLE
LDAP plugin locks out operator after restart

### ISSUE THREAD

**Issue by laferrieren** [jenkinsci/kubernetes-operator#191]
After following the instructions in https://github.com/jenkinsci/kubernetes-operator/issues/133 and getting it working for first time, the operator appears to then be locked out of the Jenkins master after then next restart.  Following these steps works for the first time the operator goes through the reconciliation loop.  A master comes up and can login with ldap credentials.  Afterwards looks like their is a race condition with the jenkins master starting and getting the config applied to be able to connect to ldap, as a result the operator is locked out.
I made sure that their is a `jenkins-operator` user and the password is in sync with the credentials in the secret.

Steps to reproduce, are 
1) do the above instructions for configuring ldap
2) kill the pod running the jenkins master for example something like `kubectl -n jenkins-test delete pod jenkins-test-cr`
3) wait for new pod to be launched
should something like this in the operator log
```
2019-11-07T15:03:05.241Z    INFO    controller-jenkins  base/reconcile.go:437   Creating a new Jenkins Master Pod jenkins-test/jenkins-jenkins-test-cr   {"cr": "jenkins-test-cr"}
2019-11-07T15:03:06.239Z    INFO    controller-jenkins  base/reconcile.go:719   Jenkins master pod starting timeout, events '[Message: Readiness probe failed: Get http://10.41.0.27:8080/login: net/http: request canceled (Client.Timeout exceeded while awaiting headers) Subobject: spec.containers{jenkins-master}]'   {"cr": "jenkins-test-cr"}
2019-11-07T15:03:06.239Z    INFO    controller-jenkins  jenkins/jenkins_controller.go:352   Setting operator version    {"cr": "jenkins-test-cr"}
```
then after wards the operator gets stuck in a loop of 
```
2019-11-07T15:03:57.356Z    INFO    controller-jenkins  base/reconcile.go:816   Generating Jenkins API token for operator   {"cr": "jenkins-test-cr"}
2019-11-07T15:03:58.385Z    WARN    controller-jenkins  jenkins/jenkins_controller.go:167   Reconcile loop failed: couldn't init Jenkins API client: Get http://jenkins-operator-http-jenkins-test-cr.jenkins-test:8080/api/json: dial tcp 10.36.31.8:8080: connect: connection refused  {"cr": "jenkins-test-cr"}
2019-11-07T15:03:58.677Z
```
At this point their is a pod running that installed the plugins defined in the cr, but isn't fully bootstrapped, as it hasn't been binned to ldap

Some other potentially useful information:
* using `v0.2.2` for the operator
* ldap plugin is 1.20

**Comment by tomaszsek** [comment]
Hi @laferrieren

The whole CR with JasC and full operator logs will help us to find the issue.
```
2019-11-07T15:03:58.385Z    WARN    controller-jenkins  jenkins/jenkins_controller.go:167   Reconcile loop failed: couldn't init Jenkins API client: Get http://jenkins-operator-http-jenkins-test-cr.jenkins-test:8080/api/json: dial tcp 10.36.31.8:8080: connect: connection refused  {"cr": "jenkins-test-cr"}
2019-11-07T15:03:58.677Z
```
This logs tells that Jenkins should be running but for some unknown reason is not(Jenkins HTTP port is closed). It can be issue with `redinessProbe` or `livenessProbe` .

Cheers

**Comment by laferrieren** [comment]
So I poked around for the networking a bit.  When the pod is re-created jenkins is listening to http, if you go directly to the pod ip address on 8080 via a browser it loads the login screen.  It completely hangs when trying to log in at that point.  It never gets to a point where its registers as an endpoint.  As a result trying to access it over a svc won't work.
So for example it looks like this
```
$ kubectl -n jenkins-test get endpoints
NAME                                            ENDPOINTS         AGE
jenkins-operator-http-jenkins-helm-kickabout    <none>            79m
jenkins-operator-slave-jenkins-helm-kickabout   <none>            79m
```

This is the CR gist:
https://gist.github.com/laferrieren/8d970c84c1c4d7fbf5ccce7ba2e1f09e
full operator logs, the starting of a new instance and applying it starts around `2019-11-07T15:00`
https://gist.github.com/laferrieren/89ba8343ab6745537c83194db7dfeb3f
JasC
https://gist.github.com/laferrieren/4ddf2a417a1b9f07cdaad67e3fb68931
also snippet of pod status from the jenkins master
https://gist.github.com/laferrieren/93d1d3035bbc61fe0d2035a7007c7fb3

Let me know if anything else would be useful

**Comment by tomaszsek** [comment]
Could you send me the results of commands `kubectl -n jenkins-test get endpoints -o yaml` and `kubectl -n jenkins-test get pod -o yaml`?

**Comment by laferrieren** [comment]
Here is output for endpoints:
https://gist.github.com/laferrieren/c487a9787a1057be82f77c3a6a29a078
pod output
https://gist.github.com/laferrieren/437b7b14861c50aa725d2d7d7368953a

**Comment by tomaszsek** [comment]
One more thing `kubectl -n jenkins-test get svc -o yaml`. I suspect that services `jenkins-operator-http-jenkins-helm-kickabout` and `jenkins-operator-slave-jenkins-helm-kickabout` have wrong selector.

**Comment by laferrieren** [comment]
here is the svc output:
https://gist.github.com/laferrieren/158bf11801e95bc61cec475c2c05bc1f

ya quickly looking at this, ya the lables don't match up, didn't think to check that as before the pod was restarted the labels where matching.  I added extra labels to the service section of the cr, is that the potential problem (they appear to have made it into the selector block)?

**Comment by tomaszsek** [comment]
I didn't manage to reproduce the issue. I've added extra labels to the `spec.service.labels` and the Jenkins http service selector didn't change. I think the Jenkins http service has been changed by hand and operator didn't fix it. I will add fix which will overwrite the service selector.

**Comment by pawelprazak** [comment]
(waiting for a release with this patch)

**Comment by BogdanSorlea** [comment]
I am running jenkins with LDAP locally just fine, the provisioning seems to work. However, there are indications in the log that the operator is trying to auth, but can't - see this issue: https://github.com/jenkinsci/kubernetes-operator/issues/208

// just in case this input helps (it's not affecting me right now, I think - I have an issue when deploying, due to annotations changing, which triggers a restart of jenkins upon reconcile, but doesn't sound related)

**Comment by tomaszsek** [comment]
Fixed in https://github.com/jenkinsci/kubernetes-operator/releases/tag/v0.3.0

**Comment by laferrieren** [comment]
Just as an update, tried this after the 0.3.0 upgrade and found what was the real root cause.  In my cr I had set which was breaking the service.  The pod had a label which was `app: jenkins-operator`, which then didn't match what the service was.  The selector for the service was using what was in the cr of `app: jenkins-master`

from my cr
```
...
  service:
    labels:
      app: jenkins-master
```
Verified this is the case from the gist's I posted before.  LDAP is working correctly after sorting out the labels.
